### PR TITLE
feature(Query): Allows query to be written to RethinkDB

### PR DESCRIPTION
**Pull request by jimberlage**
Adds the ability to send a query, as described in [writing drivers](https://rethinkdb.com/docs/writing-drivers/).  @novemberalpha any thoughts?

### CODE REVIEW

**Comment by novemberalpha** [comment]
Cool! I just installed Rust and created a "Hello World!" app per the tute. 

My first questions to you on the PR are nitty but I think an open source project merits the "ceremony." I want to make sure I'm adding value though so make sure you tell me if you don't care about this stuff right now. I'm not religious about this stuff but it presents better in open source code and prevents long threads about style later. :-)
1. Look at the commenting [here](https://github.com/rethinkdb/rethinkdb/blob/next/src/rdb_protocol/ql2.proto). It's freaking massive. The point is, however, to explain what's happening to even the dimmest coder (like me!). Are you planning on commenting this code?
2. I'm just digging into this lang but it seems like there are uses for snake_case and InitialCapsCase in the code. There are not, however usages that Mix_Cases_LikeThis. Take a look [here](https://www.rust-lang.org/en-US/faq.html#how-do-i-return-a-borrow-to-something-i-created-from-a-function). It appears local variables are represented in snake_case and InitialCaps is used for classes and whatnot. In your code you use "VersionDummy_Version" which mixes the two cases. There are only a couple of times you in your code but sticking to one case-type will help coders understand what is being described.

I'll dig in more on the actual function level performance next. I've got to run an errand so hopefully I can dig in this afternoon after I'm back.

**Comment by jimberlage** [comment]
Fully agreed on the commenting.  It's way too sparse right now.  I'll fix that and push a PR for it, and update the README too.  It seems like as good a time as ever for that.

Unfortunately, the mixed case is something we can't get away with.  [ql2.rs](https://github.com/jimberlage/rethinkdb.rs/blob/master/src/ql2.rs) (where the Term_TermType and other ugly definitions come from) is an auto-generated file based on the [ql2.proto](https://github.com/jimberlage/rethinkdb.rs/blob/master/ql2.proto) file in the repo.  I was going to replace them with aliases, actually, if we ever need to expose the definitions there publicly.  Something like:

``` rust
type TermType = Term_TermType;
```

**Comment by jimberlage** [comment]
The proto file lives here: https://github.com/rethinkdb/rethinkdb/blob/next/src/rdb_protocol/ql2.proto

I've copied it into the repo so that people know what version of the file I'm targeting.

**Comment by novemberalpha** [comment]
I literally just had a mini-stroke. Who names variables like that??!? Mayhem.

**Comment by novemberalpha** [comment]
OBSOLETE_noreply = WTHFHFHFH!!!

**Comment by novemberalpha** [comment]
That's like three case rules smashed into one.

File: src/connection.rs [hunk 1]
```diff
@@ -1,13 +1,15 @@
-use byteorder::{WriteBytesExt, LittleEndian};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
```
Comment (novemberalpha) [inline src/connection.rs:1]: I for one prefer your ordering the parameters this way. I'm sure this wasn't an ordering thing and had more to do with actual parameter order, but still def, read, write = nice.

File: src/connection.rs [hunk 2]
```diff
@@ -19,29 +21,36 @@ pub struct Connection {
...
+    length:      u32,
+    response:    Json,
 }
 
 /// The response returned by V1_0 of the RethinkDB handshake protocol, after a protocol version has
 /// been successfully set.
 #[derive(Debug,RustcDecodable)]
 struct ProtocolSuccessResponse {
-    success: bool,
+    success:              bool,
```
Comment (novemberalpha) [inline src/connection.rs]: What's going on with the spacing? Is this a monotype alignment thing that's not showing in github correctly?

File: src/connection.rs [hunk 3]
```diff
@@ -187,11 +196,12 @@ impl Connection {
     pub fn connect(host: &str, port: u16, user: &str, password: &str) -> Result<Connection, Error> {
         let stream = my_try!(TcpStream::connect((host, port)));
         let conn = Connection{
-            host:     host.to_string(),
-            port:     port,
-            stream:   RefCell::new(stream),
-            user:     user.to_owned(),
-            password: password.to_owned(),
+            host:        host.to_string(),
```
Comment (novemberalpha) [inline src/connection.rs]: Why convert this to a string()?

File: src/connection.rs [hunk 4]
```diff
@@ -207,4 +217,59 @@ impl Connection {
             },
         }
     }
+
+    fn send_query(&self, tree: &Tree) -> Result<(), Error> {
+        let token = self.query_token.get();
+        // Increment the token for the next request.
+        self.query_token.set(token.wrapping_add(1));
```
Comment (novemberalpha) [inline src/connection.rs:224]: I think there are two techniques for setting the wrapping attribute in rust. One is a function as you've done here and the other is declaring it in the definition. I prefer the second as this is required everywhere IMO. 

https://doc.rust-lang.org/1.6.0/std/num/struct.Wrapping.html

File: src/connection.rs [hunk 5]
```diff
@@ -207,4 +217,59 @@ impl Connection {
...
+        let tree = my_try!(json::encode(&tree.to_json()));
+        let len = tree.as_bytes().len();
+        if len > (u32::MAX as usize) {
+            return Err(Error::QueryTooLarge(len));
+        }
+
+        my_try!(self.stream.borrow_mut().write_u64::<LittleEndian>(token));
```
Comment (novemberalpha) [inline src/connection.rs:231]: :-)

File: src/connection.rs [hunk 6]
```diff
@@ -207,4 +217,59 @@ impl Connection {
...
+    // collection.
+    fn parse_response_from_stream(stream: &TcpStream) -> Result<QueryResponse, Error> {
+        let mut reader = BufReader::new(stream);
+        let token = my_try!(reader.read_u64::<LittleEndian>());
+        let len = my_try!(reader.read_u32::<LittleEndian>());
+        let mut recv = vec![];
+        let _ = my_try!(reader.take(len as u64).read(&mut recv));
+        let response = my_try!(Json::from_str(my_try!(str::from_utf8(&recv))));
+
+        Ok(QueryResponse {
```
Comment (novemberalpha) [inline src/connection.rs:249]: Is there a not OK return?

File: src/connection.rs [hunk 7]
```diff
@@ -207,4 +217,59 @@ impl Connection {
...
+        // Increment the token for the next request.
+        self.query_token.set(token.wrapping_add(1));
+        let tree = my_try!(json::encode(&tree.to_json()));
+        let len = tree.as_bytes().len();
+        if len > (u32::MAX as usize) {
+            return Err(Error::QueryTooLarge(len));
+        }
+
+        my_try!(self.stream.borrow_mut().write_u64::<LittleEndian>(token));
+        my_try!(self.stream.borrow_mut().write_u32::<LittleEndian>(len as u32));
```
Comment (novemberalpha) [inline src/connection.rs:232]: Is this where the my_try! macro is declared?

File: src/connection.rs [hunk 8]
```diff
@@ -207,4 +217,59 @@ impl Connection {
...
+
+        result.unwrap()
+    }
+
+    pub fn db_create(&self, name: &str) -> Result<QueryResponse, Error> {
+        my_try!(self.send_query(&Tree::Query {
+            head: Term_TermType::DB_CREATE,
+            tail: vec![Tree::Datum(Json::String(name.to_owned()))],
+        }));
+        Ok(my_try!(self.parse_response()))
```
Comment (novemberalpha) [inline src/connection.rs:273]: Error return logic?

File: src/reql/mod.rs [hunk 9]
```diff
@@ -1,2 +1,2 @@
 mod command;
-mod tree;
+pub mod tree;
```
Comment (novemberalpha) [inline src/reql/mod.rs:1]: Avoid global / public whenever possible.

File: src/reql/tree.rs [hunk 10]
```diff
@@ -1,5 +1,6 @@
+use protobuf::ProtobufEnum;
 use rustc_serialize::json::{Json, ToJson};
-use super::ql2::Term_TermType;
+use super::super::ql2::Term_TermType;
```
Comment (novemberalpha) [inline src/reql/tree.rs:3]: This is a double super class climb? Kinda getting messy.

**Review by novemberalpha — COMMENTED** [review]
I'm bringing a bunch of dumb questions and assumptions to this. Many may be just plain off but I hope these comments are helpful.

File: src/connection.rs (hunk 2 above)
Comment (jimberlage) [reply src/connection.rs]: That's off.   I should really just run rustfmt on this codebase.

File: src/connection.rs (hunk 3 above)
Comment (jimberlage) [reply src/connection.rs]: Ah.  There are two types of strings in Rust - the `&str` type, which is a reference to a slice of an existing string, and is read-only for memory safety.  There is also the `String` type, which is owned by whatever function/struct is using it and can be modified.

`to_string()` is the same as `to_owned()` - it takes an `&str`, copies the bits, and returns a new `String` which can be modified without changing the original string.  For consistency's sake I should have used `to_owned()` here, as I did with the other variables.

File: src/connection.rs (hunk 7 above)
Comment (jimberlage) [reply src/connection.rs:232]: It lives in [macros.rs](https://github.com/jimberlage/rethinkdb.rs/blob/master/src/macros.rs).  There's some comment on how it works in my [documentation PR](https://github.com/jimberlage/rethinkdb.rs/pull/11).

File: src/connection.rs (hunk 6 above)
Comment (jimberlage) [reply src/connection.rs:249]: The `my_try!` macro will return `Err` explicitly.  I could also have written this

``` rust
match Json::from_str(my_try!(str::from_utf8(&recv)) {
  Ok(response) => Ok(QueryResponse {
    query_token: token,
    length:      len,
    response:    response,
  }),
  Err(error) => Err(error),
}
```

Essentially, whenever you see the `my_try!` macro, it's doing error handling on whatever is being called and returning the error if it exists.  So if you reach the last statement in the function, it's guaranteed to be `Ok`.  I may change course though, and use an explicit match statement on the last item returned for the sake of readability.

File: src/connection.rs (hunk 8 above)
Comment (jimberlage) [reply src/connection.rs:273]: See the above comment - the `my_try!` macro handles the error case.

File: src/reql/mod.rs (hunk 9 above)
Comment (jimberlage) [reply src/reql/mod.rs:1]: In this case it's strictly a necessity - the parent module needs to access this, so it must be made public.  The parent library will not expose this module to the user of this library, though - `reql` is not public: https://github.com/jimberlage/rethinkdb.rs/blob/feature/wip/src/lib.rs#L13

File: src/reql/tree.rs (hunk 10 above)
Comment (jimberlage) [reply src/reql/tree.rs:3]: True.  I don't like this but need to figure out what the right way around it is.

File: src/connection.rs (hunk 4 above)
Comment (jimberlage) [reply src/connection.rs:224]: I'm going to keep the wrapping_add, just for simplicity.  You can wrap the type in a Wrapping struct, but there's a lot more boilerplate.   You can play around with it here:

https://play.rust-lang.org/?gist=81e3d3ac27a87b2a9ad451f774de8cbd&version=stable&backtrace=0

**Comment by jimberlage** [comment]
I appreciate the comments!

**Pull request merged**